### PR TITLE
os/bluestore: add discard method for ssd's performance

### PR DESCRIPTION
**Pull request by voidbag**
discard method is added to BlockDevice.h for ssd's performance
For ssd's performance, discard should be used, when BlueStore or BlueFS releases block device's area.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

### CODE REVIEW

**Comment by liewegas** [comment]
We should add a bdev_discard bool config option to enable/disable this, as discard is bad on some devices.

Also, teh bluestore piece is more complicated.  The existing discard op is synchronous.. we don't want to call it where you have it.  We either need a thread doing these, or to use aio.

And we probably need to add some ordering infrastructure to make sure that a deallocate+discard doesn't run concurrently with a new allocation and write into that space.  I'm not sure exactly what the rules are, but at a minimum we need to make sure the discard aio completes before allowing the space to be reallocated (i.e., discard before release to allocator).

Note that the last time I had a conversation with @allensamuels (sandisk) about this we decided it wasn't that important that we worry about discards.  I forget what the reason was.

**Comment by varadakari** [comment]
This needs a rebase.

**Comment by varadakari** [comment]
Not all the vendors implement discard/trim mechanism, they can take the request and drop/not implement in the firmware level. But it is generally good idea to pass the discards to help the gc process. As Sage mentioned we need to measure the cost of sync calls vs async, some test results if possible on a aged device with gc running.

**Comment by voidbag** [comment]
@varadakari
I'll rebase it
I had a lot of things to do...
I will update it based on Sage 's comment until next week.
Sorry for the delay.

Thanks

**Comment by voidbag** [comment]
@wido
I think i have done the implementation.
The remaining thing is to fix the failure of the test, which u and other can point out.
I'm also trying to find the cause of the failure of test.

Thanks

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -1889,7 +1918,17 @@ void BlueFS::sync_metadata()
...
+    if (cct->_conf->bdev_enable_discard && cct->_conf->bdev_async_discard) {
+      if (!to_release[i].empty())
+        r = bdev[i]->queue_discard(to_release[i]);
+      if (r == 0)
+        continue;
+    }
+
     for (auto p = to_release[i].begin(); p != to_release[i].end(); ++p) {
+      if (cct->_conf->bdev_enable_discard && r == 0)
+        bdev[i]->discard(p.get_start(), p.get_len());
```
Comment (liewegas) [inline src/os/bluestore/BlueFS.cc]: This doesn't seem right.. first, if you discard async, you do'nt want ot later discard() sync, or call release(), right?

**Comment by wido** [comment]
So this one is still pending?

**Comment by liewegas** [comment]
Yeah, I'm still nervous about it. Two things:

1. The special async interface for discard, separate from teh aio one, 
rubs me the wrong way, although it's probaby justified since you can't 
submit discard requests via libaio

2. I'm not sure this will actually be a good thing to do on a real SSD.  
For XFS most seem to recommend using fstrim periodically instead of online 
discard because so many firmwares suck and there are often performance 
anomalies (or bugs) from lots of small discards.  Even if that isn't the 
case, if we are discarding small bits as we free them, they may be dropped 
by teh SSD such that we don't trim the larger region they eventuallly 
form.

I suppose we should probably rebase this, and do some testing, but leave 
it off by default (maybe with the option marked DEV or something so that 
there is some warning for users?)?

**Comment by voidbag** [comment]
I'll rebase this pr and check the issues you all pointed out, until this weekend

@liewegas I think fstrim is for just local filesystem unlike BlueFS using raw block device.
So i think BlueFS has to issue discard command to raw block device like other filesystem.

**Comment by liewegas** [comment]
Right about fstrim.  I mean that it is a periodic operation that reexamines the freelist and trims it, versus an online approach that trims as we free things.  The offline one will submit fewer trims and they should (as well as the device allows) regardless of the size of the individual released extents.

**Comment by liewegas** [comment]
FWIW I think we will want to have an fstrim-like function either way in order to discard upgraded bluestores.  One it's there we can compare results of a periodic approach vs the online one.  I think the StupidAllocot in-memory structure should make it reasonably easy to implement since it already orders/batches free extents by size; we can focus on discarding the big extents first (or only).

**Comment by wido** [comment]
@liewegas The problem with mounting EXT4 or XFS with the *discard* option isn't that discarding is the problem, but it's SATA 3.0

The SATA 3.1 spec brings queued trimming which overcomes these problems. Many controllers out there however only implement SATA 3.0.

A few resources:
- http://techreport.com/news/21311/sata-3-1-spec-brings-swap-standard-improved-trim
- https://en.wikipedia.org/wiki/Serial_ATA#SATA_revision_3.1

So yes, off by default and let the user enable it if the need to. Might need to write a log line telling them that in certain scenarios having discard on can degrade performance.

File: src/os/bluestore/BlueFS.cc (hunk 1 above)
Comment (voidbag) [inline src/os/bluestore/BlueFS.cc]: I think this code is also wrong, because r is undefined in that scope...
The intention of the code is that it discards given range which is failed to discard async, synchronously.

And i don't understand your(@liewegas) comment "This doesn't seem right.. first, if you discard async, you do'nt want ot later discard() sync, or call release(), right?"
Does it mean 'Don't discard synchronously and don't call release(), if discard has been done async'
Am i right??

**Comment by voidbag** [comment]
@liewegas 
I read all your comments.
I'll briefly describe what I have to do according to your comments.
Step1: Rebase this, fix bugs and add options whose postfix is _dev (This PR)
Step2: Implement periodic discard comment using stupid alloc with only big extents. (new PR)
Step3: Implement fstrim function with every extents like other filesystem (xfs..) (new PR)

Am I right? Plz correct me if i made a mistake.

File: src/os/bluestore/BlueFS.cc (hunk 1 above)
Comment (liewegas) [reply src/os/bluestore/BlueFS.cc]: That's a comment I deleted after I looked harder and understood the code :)

**Comment by liewegas** [comment]
I think 3 isn't needed; just 2!

**Comment by voidbag** [comment]
Okay, then,
First, i tried to add _dev options in this PR

File: src/common/config_opts.h [hunk 2]
```diff
@@ -1024,6 +1024,8 @@ OPTION(bdev_debug_aio_suicide_timeout, OPT_FLOAT, 60.0)
 // NVMe driver is loaded while osd is running.
 OPTION(bdev_nvme_unbind_from_kernel, OPT_BOOL, false)
 OPTION(bdev_nvme_retry_count, OPT_INT, -1) // -1 means by default which is 4
+OPTION(bdev_enable_discard, OPT_BOOL, true)
```
Comment (wido) [inline src/common/config_opts.h]: Shouldn't these be set to false?

**Comment by wido** [comment]
Do we have an update on this PR?

There are more SATA 3.1 and SAS 3.0 SSDs (Samsung PM1633a) out there which async queue TRIM so it can be 'sync' in the code.

Would like to see this in BlueStore

**Comment by voidbag** [comment]
Rebase has been done

**Comment by voidbag** [comment]
Can you all review this code?

**Comment by voidbag** [comment]
Before being reviewed, I'll try to fix the seg fault..
Thanks

**Comment by liewegas** [comment]
This looks quite reasonable!  Ping me when you've tracked down the seg fault.

**Comment by voidbag** [comment]
The reason of seg fault was the access to freed allocator in static callback.
So, I added drain method of discard thread to resolve this problem.

Can you all review this code?
Thanks.

File: src/os/bluestore/KernelDevice.cc [hunk 3]
```diff
@@ -440,6 +487,55 @@ void KernelDevice::_aio_thread()
...
+      dout(20) << __func__ << " finishing" << dendl;
+      for (auto p = discard_finishing.begin();p != discard_finishing.end(); ++p) {
+	discard(p.get_start(), p.get_len());
+      }
+
+      discard_callback(discard_callback_priv, static_cast<void*>(&discard_finishing));
+      l2.lock();
+      discard_finishing.clear();
+      l2.unlock();
+      l.lock();
```
Comment (voidbag) [inline src/os/bluestore/KernelDevice.cc]: Can I access discard_finishing interval_set without mutex?
(l.unlock() -> accessing discard_finishing -> l.lock())
At first, i didn't use l2.lock, because discard_finishing.clear() is between l.unlock() and l.lock(), which can act like memory fence.
Nevertheless, I inserted l2 lock to protect discard_finishing interval set for safety.

I didn't use the way to put discard_finishing.clear() after l.lock(), because discard_finishing.clear() can block discard queuing thread.

Thanks

Comment (liewegas) [reply src/os/bluestore/KernelDevice.cc]: Why not use a single lock, and include a bool discard_running; next to discard_submitted.  Then you can 

while (!discard_submitted.empty() || discard_running) {
   wait
}

under that single lock for drain?

Comment (voidbag) [reply src/os/bluestore/KernelDevice.cc]: I understand the suggestion.
With atomic variable "discard_running",
I2 lock can be removed.

I'll update the patch!

Comment (liewegas) [reply src/os/bluestore/KernelDevice.cc]: Not sure it even needs to be atomic.. teh drain method has to take the lock anyway to check discard_submitted, so it can be a regular bool?

Comment (voidbag) [reply src/os/bluestore/KernelDevice.cc]: Aha, the running variable is accessed within lock scope! Then, I think it doesn't have to be atomic variable as you said.

I'll rewrite that part as soon as possible. Isn't there any other suggestion for this PR?

Comment (liewegas) [reply src/os/bluestore/KernelDevice.cc]: Nope!

I think as a follow-up we'll want to do periodic bulk discard, but that's independent of this.

File: src/common/options.cc [hunk 4]
```diff
@@ -3169,6 +3169,14 @@ std::vector<Option> get_global_options() {
...
     .set_description(""),
 
+    Option("bdev_enable_discard", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
+    .set_default(true)
+    .set_description(""),
+
+    Option("bdev_async_discard", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
+    .set_default(true)
+    .set_description(""),
+
```
Comment (wido) [inline src/common/options.cc]: Should we set both options to true by default? Don't we want to disable this by default and let users enable it if they want to?

Comment (voidbag) [reply src/common/options.cc]: @wido 
I set the flag as true, for checking the correctness of this option in jenkins test.
I agree with you.
In addition to changing the default value, some variables' names will be also modified.
Thanks for reviewing it.

File: src/os/bluestore/KernelDevice.cc (hunk 3 above)
Comment (voidbag) [reply src/os/bluestore/KernelDevice.cc]: Okay. I'll implement that part with stupid allocator first, using coarse-grained lock as soon as possible. With BitmapAllocator, fine-grained locking mechanism will be introduced.

I think the background(periodic) thread will be implemented like sync_thread in FileStore.
Thanks.

File: src/common/options.cc (hunk 4 above)
Comment (wido) [reply src/common/options.cc]: Ok, understood. Just checking :)

We might want to prefix these with *bluestore* indeed as they apply to BlueStore only.

What do you say @liewegas ?

Comment (voidbag) [reply src/common/options.cc]: I just thought this feature was for bdev class, specifically.
Reasonable suggestion! It's possible to change the option name as you said.
:)

**Comment by myoungwon** [comment]
Async discard does not free extents immediately. Therefore, in the worst case, multiple block allocation can be blocked if discard thread (a single thread) is not fast enough. 
This seems to have a negative impact on block allocation strategy or performance.

How about adding a function that discards and release extents before allocating blocks if discard queue has a large number of entries.

**Comment by liewegas** [comment]
I wouldn't bother with BitmapAllocator--just do StupidAllocator.

I think what we want is a call that returns extents larger than size X after offset Y.  Then you can just allocate the extent, discard it, and once it discards release it back to the allocator.  I think the only real thing to watch out for is to not allocate all the space accidentally--i.e., never return an extent that is more than 1/2 of the freespace or something.

**Comment by voidbag** [comment]
@myoungwon 
I'll introduce throttling mechanism as you said.
Thanks for your comment

@liewegas 
I didn't understand what you said exactly.
Does "never return an extent that is more than 1/2 of the freespace or something." mean that allocator cannot allocate large extent considering current freespace? I didn't understand the purpose of this constraint... Could you explain in detail?

Thanks

**Comment by voidbag** [comment]
Drain has been modified with single lock, using discard_running

Remaining job for this pr:
1. Throttling mechanism will be introduced, as @myoungwon suggested.
2. option name and default value will be changed, as @wido suggested.

Thanks.

**Comment by liewegas** [comment]
@voidbag I'm just thinking ahead to how the bulk discard shoudl work.  I'm guessing that the simplest thing to avoid races (we decide to bulk discard some large reason, start discard; a write comes along and allocates part of that region and writes to it; write and discard race and tears follow), we should "allocate" the regions we are discarding (to avoid them being used) and then release them again once the discard operation completes.  In that case, we'd just need to make sure we don't allocate the entire freespace (for a mostly-empty device with one huge free extent) such that writes get ENOSPC.

**Comment by liewegas** [comment]
For the throttling, I would be inclined to make it fall back to not doing discard instead of doing a synchronous discard.  I am very nervous about devices with a very slow discard implementation.  Am I being to paranoid/conservative?  It could be an option, I guess (fall back to sync discard or fall back to no discard).  

Honestly I suspect that the thing we'll want to recommend is periodic bulk discard, not this fine-grained online discard.

**Comment by voidbag** [comment]
To @liewegas 

- About bulk discard
I think bulk discard has to be implemented by the sequence ('allocate, discard, release')  as you mentioned. And I finally understand your intention that we should not allocate too much region for "bulk discard", otherwise future ongoing write cannot be allocated in case of full allocation caused by discard.
To implement it, I think we have to manage free-list with aging. I completely understand what you mean
Thank you for your detailed description!

- About throttling
I agree with you. I think it's better not to discard, in case of write-intensive situation.
I'll implement fallback to no discard, instead of throttling. @myoungwon

- About option name.
What about changing the prefix with "bluestore" as @wido mentioned.

**Comment by liewegas** [comment]
Yeah, bluestore_ prefix sounds good.

Thanks!!

**Comment by myoungwon** [comment]
It sounds good to me. I asked a few questions about TRIM command to our firmware engineer (SK Hynix). The real benefit of TRIM is that it can reduce copy operation during gc (garbage collection) processing in SSD. So, if we can mark a page (in SSD, 4KB aligned) as "unused", gc can be faster (because marked page does not need to be copied).

But, TRIM needs to update metadata in SSD (for example, set invalid bit in FTL (flash translation layer)).
So, burst TRIM commands can cause the performance degradation as @liewegas comments.

In conclusion, I think no TRIM can be answer if we need to send a lot of commands to preserve the performance.

**Comment by wido** [comment]
@liewegas: If we disable TRIM by default nobody would be impacted by performance problems, right?

Yes, there are crappy SSDs out there with bad TRIM.

But with proper SATA 3.1 controllers and SSDs or SAS you will be fine.

On Intel and Samsung DC-grade SSDs I don't see major problems with TRIM. It usually the controller in between which does weird things.

File: src/os/bluestore/BlueFS.cc [hunk 5]
```diff
@@ -222,6 +242,13 @@ int BlueFS::reclaim_blocks(unsigned id, uint64_t want,
   return 0;
 }
 
+void BlueFS::handle_discard(unsigned dev, interval_set<uint64_t>& to_release)
+{
+  dout(10) << __func__ << dendl;
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: May be output dev here as well?

**Comment by wido** [comment]
@myoungwon Could you resolve the conflict?

**Comment by myoungwon** [comment]
@voidbag will resolve the conflict within a day. I told him to resolve this problem.

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -8335,9 +8352,20 @@ void BlueStore::_txc_release_alloc(TransContext *txc)
     // bstore_kv_final. We could serialize in simpler format and perform
     // the merge separately, maybe even in a dedicated thread.
     bulk_release_extents.insert(txc->released);
+    int r = 0;
+    if (cct->_conf->bdev_enable_discard && cct->_conf->bdev_async_discard) {
+      r = bdev->queue_discard(bulk_release_extents);
+      if (r == 0)
+	goto out;
```
Comment (myoungwon) [inline src/os/bluestore/BlueStore.cc]: @voidbag I think we need to invoke alloc->release() even though bdev->queue_discard is fail?

**Comment by wido** [comment]
Ah, my bad @myoungwon , I meant @voidbag :)

**Comment by voidbag** [comment]
@wido 
I think discard feature is related to only block device.
So, I think the prefix of the discard feature should be bdev, rather than bluestore.
What about your opinion??

File: src/os/bluestore/BlueFS.cc (hunk 5 above)
Comment (voidbag) [reply src/os/bluestore/BlueFS.cc]: @ifed01 
I don't understand what you said.
Could you say more concretely?

**Comment by wido** [comment]
@voidbag Fair point! That does seem consistent with other values, right? They also start with bdev.

Don't forget to make the defaults go to false ;)

File: src/os/bluestore/BlueFS.cc (hunk 5 above)
Comment (ifed01) [reply src/os/bluestore/BlueFS.cc]: I just meant that you can additionally print value of 'dev' var here.

Comment (voidbag) [reply src/os/bluestore/BlueFS.cc]: I updated the log message.
Thanks for your comment!

**Comment by voidbag** [comment]
@wido 
I won't forget to do it.
I will change the default value as false, when this patch is about to be merged.
Thanks

**Comment by wido** [comment]
How are we with this PR? @voidbag 

Thanks!

**Comment by liewegas** [comment]
Queuing this for testing.  I think there are just 2 changes needed:

1. change default to false
2. update qa/objectstore/bluestore*.yaml so that we have variations that enable discard.  I think we probably need to test compression w/ and without discard.  Maybe have it off (default) for the bluestore-bitmap.yaml but enable it for bluestore.yaml?

**Comment by wido** [comment]
Ok! @voidbag could you set the defaults to False?

**Comment by voidbag** [comment]
@wido @liewegas 
Default value has been changed to false.
Should I also update qa/objectstore/bluestore*.yaml?
Thanks

**Comment by liewegas** [comment]
Yes, please!

**Comment by wido** [comment]
@voidbag Have you been able to modify these values?

That should be the last thing before this PR can be merged

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-01-11_05:03:22-rados-wip-kefu-testing-2018-01-11-1036-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-01-14_09:26:29-rados-wip-kefu-testing-2018-01-11-1036-distro-basic-mira/

failed tests are tracked by 

- http://tracker.ceph.com/issues/20236
- http://tracker.ceph.com/issues/22527
- http://tracker.ceph.com/issues/22668

**Comment by k0ste** [comment]
I'm properly understand that if `bdev_enable_discard` is enabled TRIM command issued for: `wal`, `db` and `db_slow` (backend) device. What if backend device is HDD?

**Comment by wido** [comment]
@k0ste Yes, it does. I think we can skip (or make it configurable) TRIM for db_slow as that's usually on a HDD or non-TRIM device.

**Comment by k0ste** [comment]
@wido yes we should. I think `is_rotational` function can be used for it.
@voidbag please take a look. TRIM should not be issued if backend device is HDD.

**Comment by wido** [comment]
@k0ste Good suggestion! @voidbag can you look into that?

File: src/os/bluestore/KernelDevice.cc [hunk 7]
```diff
@@ -659,6 +748,14 @@ int KernelDevice::aio_write(
   return 0;
 }
 
+int KernelDevice::discard(uint64_t offset, uint64_t len)
+{
+  int r = 0;
+  if (!rotational)
+    r = block_device_discard(fd_direct, (int64_t)offset, (int64_t)len);
+  return r;
+}
```
Comment (voidbag) [inline src/os/bluestore/KernelDevice.cc]: @wido @k0ste 
I had already implemented that part.
You all don't have to worry about issuing discard command to rotational device.

**Comment by wido** [comment]
I see @voidbag ! You are correct

**Comment by wido** [comment]
@voidbag Can you resolve the conflicts?

After that I hope it's ready to be merged

**Comment by voidbag** [comment]
@wido Rebase has been done.

Could anyone tell me how the bluestore*.yaml file has to be modified?
Do  I have to modify existing file or create new one. 
If I have to create new one, where do I have to do?

Periodic discard
I have analyze allocator code(Stupid, Bitmap), roughly.
And I realized that bitmap allocator has fixed range of disk unlike stupid allocator that has only free area.
There isn't any good idea with bitmap allocator in my mind.
Current bad idea is a flag for undiscarded block, which requires full search among discarded and undiscarded blocks.
So,
I'll start to write the periodic discard separating undiscarded and discarded blocks with 'stupid allocator', as soon as this pr is merged.

Thanks

**Comment by wido** [comment]
Thanks! @liewegas might know better, but I think that *./qa/objectstore/bluestore.yaml* is sufficient.

Although the testing HW should also support discard in order to make this work.

Would be very nice if this one gets merged. It has been open for a very long time.

**Comment by liewegas** [comment]
Just enable the setting in bluestore.yaml.. that should be sufficient.  Thanks!

**Comment by liewegas** [comment]
The smithi testing hardware that we use primarily for rados is all nvme based so this should be exercised.

**Comment by wido** [comment]
Great, hopefully the tests are OK so that this one can be merged.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-02-23_01:06:09-rados-wip-kefu-testing-2018-02-23-0044-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-02-24_00:20:51-rados-wip-kefu-testing-2018-02-23-0044-distro-basic-mira/

the failures in the first run were caused by sepia issues.

File: src/os/bluestore/BlueStore.cc [hunk 8]
```diff
@@ -8286,10 +8303,25 @@ void BlueStore::_txc_release_alloc(TransContext *txc)
 {
   // it's expected we're called with lazy_release_lock already taken!
   if (likely(!cct->_conf->bluestore_debug_no_reuse_blocks)) {
-    dout(10) << __func__ << " " << txc << " " << std::hex
+    int r = 0;
+    if (cct->_conf->bdev_enable_discard && cct->_conf->bdev_async_discard) {
+      r = bdev->queue_discard(txc->released);
```
Comment (interestingyong) [inline src/os/bluestore/BlueStore.cc:8308]: if osr->q  contains write txc
did discard operation will later than DIO operation ?
Data will be lost ?

File: src/os/bluestore/BlueStore.cc [hunk 9]
```diff
@@ -8286,10 +8303,25 @@ void BlueStore::_txc_release_alloc(TransContext *txc)
...
+    } else if (cct->_conf->bdev_enable_discard) {
+      for (auto p = txc->released.begin(); p != txc->released.end(); ++p) {
+	  bdev->discard(p.get_start(), p.get_len());
+      }
+    }
+    dout(10) << __func__ << "(sync) " << txc << " " << std::hex
              << txc->released << std::dec << dendl;
     alloc->release(txc->released);
   }
+
```
Comment (interestingyong) [inline src/os/bluestore/BlueStore.cc:8323]: if osr->q contains write txc
did discard operation will later than DIO operation ?
Data will be lost ?
Should  check  that txc op as delete /trucate type  ?
